### PR TITLE
Add initial set of conformance tests for Stream

### PR DESCRIPTION
**Pull request by stephentoub**
This also includes fixes to a variety of streams where they're out of conformance, in particular around argument names.

The tests being added are primarily focused on "connected streams", ones where two streams are used to represent two ends of a communication channel.  We can subsequently flesh this out for other kinds of streams.  Also, while I consolidated a bunch of tests that were spread across different stream-derived type tests, there's still more that can be consolidated over time into these centralized tests.

cc: @geoffkizer, @ericstj, @terrajobst

### CODE REVIEW

File: src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs:28]: We should instance this for multiple providers, like the other QUIC tests

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs:11]: Why is this called SslStream**Memory**ConformanceTests?

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs:17]: Why doesn't this use the Wrapped version of the conformance tests?

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs:284]: Might be worth keeping this just for DataAvailable

**Review by geoffkizer — APPROVED** [review]
In general this looks really good. A few minor comments below.

Are you planning additional work here, or is this basically done?

**Comment by geoffkizer** [comment]
I think we should also consider using the WrappedConnectedStreamTests for HttpClient request/response streams. This is a little weird because we don't actually have a Stream implementation on the server side -- but we could hack this together on top of the various loopback server stuff fairly easily. Thoughts?

File: src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs
Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:543]: In many of the tests I actually want to do both directions with the same instances, in particular for bidirectional streams, such that we write to one and read from the other and then we swap.

In cases where only one direction is needed, it just looks at one.

I think the current set up is reasonable, with minimal ceremony.  If we can come up with something better subsequently, we can revisit.

File: src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs
Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:796]: Yup, that's a holdover from before I added it.

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs
Comment (stephentoub) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs:11]: It uses an in-memory channel.  I was thinking, given typical use cases, we'd want to also add a similar set over NetworkStream.

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs
Comment (stephentoub) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamConformanceTests.cs:17]: It should.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs:284]: Ok

**Comment by stephentoub** [comment]
> I think we should also consider using the WrappedConnectedStreamTests for HttpClient request/response streams. This is a little weird because we don't actually have a Stream implementation on the server side -- but we could hack this together on top of the various loopback server stuff fairly easily. Thoughts?

There are more streams in dotnet/runtime than I've included here.  My thinking was we'd get this framework in, and then folks can add additional streams to test and additional tests over time.  I got many of the heavy hitters here.

File: src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs
Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:286]: Lots of stream don't throw from flush, treating it entirely as a nop.  I could add a virtual property specifically about it; just didn't seem worth it.

File: src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs
Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:265]: I can try to improve it.  The problem is when the Can methods return false, all bets are somewhat off, and there's a fair amount of variation that happens here.

**Comment by geoffkizer** [comment]
@scalablecory Can we go ahead and fix the Dispose behavior in QuicStream? Any reason to hold off on this? There are other issues in #756 that may need additional consideration, but it would be nice to fix the Dispose issues and make these conformance tests work for QuicStream.

**Comment by scalablecory** [comment]
> @scalablecory Can we go ahead and fix the Dispose behavior in QuicStream? Any reason to hold off on this? There are other issues in #756 that may need additional consideration, but it would be nice to fix the Dispose issues and make these conformance tests work for QuicStream.

No reason to hold off. It has been a thorn in my side for HTTP/3 so I would very much like to fix it :).

File: src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs
Comment (stephentoub) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs:20]: @geoffkizer, fyi, my machine isn't set up to run these, and they don't appear to run in CI yet (correct me if I'm wrong). So you may need to tweak them once they're merged.

**Comment by gewarren** [comment]
@stephentoub @PriyaPurkayastha Can you create a breaking change docs issue for this [here](https://github.com/dotnet/docs/issues/new?assignees=&labels=&template=dotnet-breaking-change.md&title=)? I believe this went out in Preview 1.

**Comment by stephentoub** [comment]
@gewarren, I've opened https://github.com/dotnet/docs/issues/23102. Thanks.

**Comment by stephentoub** [comment]
@ericstj, you assigned this to me.  What's the action item?

**Comment by ericstj** [comment]
Mail is coming shortly.  I'm going through all PRs with needs-breaking-change-doc and asking folks to file those issues and remove the label.

**Comment by stephentoub** [comment]
Ok.  As noted above in the comments, https://github.com/dotnet/docs/issues/23102 was already opened and addressed.

**Comment by ericstj** [comment]
Thanks for removing the label, that will remove this from my queries.

**Pull request merged**
